spyderlib/widgets/externalshell/baseshell.py

```diff
@@ -290,10 +293,29 @@ def get_stderr(self):
         return self.transcode(qba)
 
     def write_output(self):
+        # if we are already writing something else,
+        # store the present message in a buffer
+        if not self.write_lock.tryLock():
+            self.buffer_lock.lock()
+            self.buffer.append(self.get_stdout())
+            self.buffer_lock.unlock()
+
+            return
+
         self.shell.write(self.get_stdout(), flush=True)
-        # Commenting the line below improves crashes on long
-        # output. See Issue 2251
-        # QApplication.processEvents()
+
+        while True:
+            self.buffer_lock.lock()
+            messages = self.buffer
+            self.buffer = []
+            self.buffer_lock.unlock()
+
+            if not len(messages): break
+            else:
+                for msg in messages:
+                    self.shell.write(msg, flush=True)                

```

Comment:
It would be more efficient to join the list first, then print. It could be harder to do with Queues however.